soc/xtensa/intel_adsp/ace_v1x/multiprocessing.c

```diff
@@ -0,0 +1,205 @@
+#define CORE_POWER_CHECK_DELAY 256
+
+/**
+ * @brief Power down procedure.
+ *
+ * Locks its code in L1 cache and shuts down memories.
+ * NOTE: there's no return from this function.
+ *
+ * @param disable_lpsram        flag if LPSRAM is to be disabled (whole)
+ * @param hpsram_pg_mask pointer to memory segments power gating mask
+ * (each bit corresponds to one ebb)
+ * @param response_to_ipc       flag if ipc response should be send during power down
+ */
+extern void ace_power_down(bool disable_lpsram, uint32_t *hpsram_pg_mask, bool response_to_ipc);
+
+#define SRAM_ALIAS_BASE         0xA0000000
+#define SRAM_ALIAS_MASK         0xF0000000
+
+#define uncache_to_cache(address) \
+				((__typeof__(address))(((uint32_t)(address) &  \
+				~SRAM_ALIAS_MASK) | SRAM_ALIAS_BASE))
+
+static ALWAYS_INLINE uint32_t processor_id_reg(void)
+{
+	uint32_t prid;
+
+	__asm__ volatile("rsr %0, PRID" : "=r"(prid));
+	return prid;
+}
+
+static void ipc_isr(void *arg)
+{
+	MTL_P2P_IPC[processor_id_reg()].agents[0].ipc.tdr = BIT(31); /* clear BUSY bit */
+#ifdef CONFIG_SMP
+	void z_sched_ipi(void);
+	z_sched_ipi();
+#endif
+}
+
+void soc_mp_init(void)
+{
+	IRQ_CONNECT(MTL_IRQ_TO_ZEPHYR(MTL_INTL_IDCA), 0, ipc_isr, 0, 0);
+
+	irq_enable(MTL_IRQ_TO_ZEPHYR(MTL_INTL_IDCA));
+
+	for (int i = 0; i < CONFIG_MP_NUM_CPUS; i++) {
+		/* DINT has one bit per IPC, unmask only IPC "Ax" on core "x" */
+		MTL_DINT[i].ie[MTL_INTL_IDCA] = BIT(i);
+
+		/* Agent A should signal only BUSY interrupts */
+		MTL_P2P_IPC[i].agents[0].ipc.ctl = BIT(0); /* IPCTBIE */
+	}
+
+	/* Set the core 0 active */
+	soc_cpus_active[0] = true;
+}
+
+void soc_start_core(int cpu_num)
+{
+	int retry = CORE_POWER_CHECK_NUM;
+	if (cpu_num > 0) {
+		/* Initialize the ROM jump address */
+		uint32_t* rom_jump_vector = (uint32_t *) ROM_JUMP_ADDR;
+		*rom_jump_vector = (uint32_t) z_soc_mp_asm_entry;
+		z_xtensa_cache_flush(rom_jump_vector, sizeof(*rom_jump_vector));
+
+		/* Tell the ACE ROM that it should use secondary core flow */
+		MTL_PWRBOOT.bootctl[cpu_num].battr |= MTL_PWRBOOT_BATTR_LPSCTL_BATTR_SLAVE_CORE;
+	}
+
+	MTL_PWRBOOT.capctl[cpu_num].ctl |= MTL_PWRBOOT_CTL_SPA;
+
+	/* Waiting for power up */
+	while (~(MTL_PWRBOOT.capctl[cpu_num].ctl & MTL_PWRBOOT_CTL_CPA) && --retry) {
+		 k_busy_wait(CORE_POWER_CHECK_DELAY);
+	}
+
+	if(!retry) {
+		__ASSERT(false, "soc_start_core() secondary core has not powered up");
+	}
+}
+
+void soc_mp_startup(uint32_t cpu)
+{
+	/* Must have this enabled always */
+	z_xtensa_irq_enable(ACE_INTC_IRQ);
+
+	/* Prevent idle from powering us off */
+	MTL_PWRBOOT.bootctl[cpu].bctl |=
+		MTL_PWRBOOT_BCTL_WAITIPCG | MTL_PWRBOOT_BCTL_WAITIPPG;
+}
+
+void arch_sched_ipi(void)
+{
+	uint32_t curr = processor_id_reg();
+
+	/* Signal agent B[n] to cause an interrupt from agent A[n] */
+	for (int core = 0; core < CONFIG_MP_NUM_CPUS; core++) {
+		if (core != curr && soc_cpus_active[core]) {
+			MTL_P2P_IPC[core].agents[1].ipc.idr = CAVS_IPC_BUSY;
+		}
+	}
+}
+
+int soc_adsp_halt_cpu(int id)
```

Comment:
@andyross I talked about this with @ceolin. I also asked You in April about it, when the initial implementation was made.  

I considered renaming so that it would not be the same function as in cAVS. But at the moment i'm more in favourite of changing cAVS implementation. For me it's misunderstanding when you are intending to halt core but in reality only allowing HW to PG in idle.